### PR TITLE
MINOR: Gracefully handle non-assigned case in fetcher metric

### PR DESCRIPTION
**Pull request by vvcephei**
Minor tweak to gracefully handle a possible IllegalStateException while checking a metric value.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by vvcephei** [comment]
I ran into this while running benchmarks. The stacktrace was:

```
Exception in thread "main" java.lang.IllegalStateException: No current assignment for partition benchmark-store-changelog-0
	at org.apache.kafka.clients.consumer.internals.SubscriptionState.assignedState(SubscriptionState.java:350)
	at org.apache.kafka.clients.consumer.internals.SubscriptionState.preferredReadReplica(SubscriptionState.java:527)
	at org.apache.kafka.clients.consumer.internals.Fetcher$FetchManagerMetrics.lambda$maybeUpdateAssignment$0(Fetcher.java:1774)
	at org.apache.kafka.common.metrics.KafkaMetric.metricValue(KafkaMetric.java:68)
```

It's the first time I hit it, so it may not be a big deal, but at the same time, it _seems_ fine to just handle this case...

**Comment by vvcephei** [comment]
Jenkins is running into OOMEs; Rather than hammer it right now, I'm just going to wait for reviews before running it again.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java [hunk 1]
```diff
@@ -1742,8 +1742,15 @@ private void maybeUpdateAssignment(SubscriptionState subscription) {
                     if (!this.assignedPartitions.contains(tp)) {
                         MetricName metricName = partitionPreferredReadReplicaMetricName(tp);
                         if (metrics.metric(metricName) == null) {
-                            metrics.addMetric(metricName, (Gauge<Integer>) (config, now) ->
-                                subscription.preferredReadReplica(tp, 0L).orElse(-1));
+                            metrics.addMetric(metricName, (Gauge<Integer>) (config, now) -> {
+                                try {
+                                    return subscription.preferredReadReplica(tp, 0L).orElse(-1);
+                                } catch (IllegalStateException e) {
```
Comment (hachikuji) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java]: Perhaps it's reasonable in this case to let `preferredReadReplica` use `assignedStateOrNull`?

**Review by hachikuji — COMMENTED** [review]
Thanks @vvcephei . Left a minor suggestion.

**Comment by vvcephei** [comment]
Hey @hachikuji , thanks for the review! I've updated the PR with what I think you meant. Please let me know if I misunderstood.

Thanks,
-John

**Review by guozhangwang — COMMENTED** [review]
The change LGTM. Thanks @vvcephei

**Review by hachikuji — APPROVED** [review]
LGTM. Thanks @vvcephei

**Comment by hachikuji** [comment]
retest this please

**Pull request merged**